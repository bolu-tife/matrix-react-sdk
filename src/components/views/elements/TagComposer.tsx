/*
Copyright 2021 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React, { ChangeEvent, FormEvent } from "react";

import Field from "./Field";
import { IInputValidationProps } from "./IInputValidation";
import { _t } from "../../../languageHandler";
import AccessibleButton from "./AccessibleButton";
import { IFieldState, IValidationResult } from "./Validation";

interface IProps extends IInputValidationProps{
    tags: string[];
    onAdd: (tag: string) => void;
    onRemove: (tag: string) => void;
    disabled?: boolean;
    label?: string;
    placeholder?: string;
}

interface IState {
    newTag: string;
    isNewTagValid: boolean;
}

/**
 * A simple, controlled, composer for entering string tags. Contains a simple
 * input, add button, and per-tag remove button.
 */
export default class TagComposer extends React.PureComponent<IProps, IState> {
    public constructor(props: IProps) {
        super(props);

        this.state = {
            newTag: "",
            isNewTagValid: false,
        };
    }

    private onInputChange = (ev: ChangeEvent<HTMLInputElement>) => {
        this.setState({ newTag: ev.target.value });
    };

    private onAdd = (ev: FormEvent) => {
        ev.preventDefault();
        if (!this.state.newTag || !this.state.isNewTagValid) return;

        this.props.onAdd(this.state.newTag);
        this.setState({ newTag: "" });
    };

    private onRemove(tag: string) {
        // We probably don't need to proxy this, but for
        // sanity of `this` we'll do so anyways.
        this.props.onRemove(tag);
    }

    private onValidateKeyword = async (fieldState: IFieldState): Promise<IValidationResult> => {
        if (!this.props.onValidate) {
            this.setState({ isNewTagValid: true });
            return { valid: false };
        }
        const result = await this.props.onValidate(fieldState);
        this.setState({ isNewTagValid: result.valid });
        return result;
    };

    public render() {
        return <div className='mx_TagComposer'>
            <form className='mx_TagComposer_input' onSubmit={this.onAdd}>
                <Field
                    value={this.state.newTag}
                    onChange={this.onInputChange}
                    label={this.props.label || _t("Keyword")}
                    placeholder={this.props.placeholder || _t("New keyword")}
                    disabled={this.props.disabled}
                    autoComplete="off"
                    onValidate={this.onValidateKeyword}
                    validateOnBlur={this.props.validateOnBlur}
                    validateOnChange={this.props.validateOnChange}
                    validateOnFocus={this.props.validateOnFocus}
                    forceValidity={this.props.forceValidity}
                />
                <AccessibleButton onClick={this.onAdd} kind='primary' disabled={this.props.disabled || !this.state.isNewTagValid}>
                    { _t("Add") }
                </AccessibleButton>
            </form>
            <div className='mx_TagComposer_tags'>
                { this.props.tags.map((t, i) => (<div className='mx_TagComposer_tag' key={i}>
                    <span>{ t }</span>
                    <AccessibleButton onClick={this.onRemove.bind(this, t)} disabled={this.props.disabled} />
                </div>)) }
            </div>
        </div>;
    }
}
