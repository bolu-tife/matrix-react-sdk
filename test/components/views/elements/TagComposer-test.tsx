import React from 'react';
import {
    act,
    renderIntoDocument,
    findRenderedDOMComponentWithTag,
    Simulate,
    scryRenderedDOMComponentsWithClass,
} from 'react-dom/test-utils';

import TagComposer from "../../../../src/components/views/elements/TagComposer";

describe("<TagComposer />", () => {
    const defaultProps = {
        tags: [],
        onAdd: jest.fn(),
        onRemove: jest.fn(),
    };
    const getComponent = (props = {}) => renderIntoDocument<React.Component>(
        <TagComposer {...defaultProps} {...props} />,
    ) as React.Component;

    const getForm = container => findRenderedDOMComponentWithTag(container, 'form');
    const getInput = container => {
        const form = getForm(container);
        return form.querySelector('input');
    };

    const getAddButton = container => {
        const form = getForm(container);
        return form.querySelector('[role=button]');
    };

    const editNewTag = async (container, value: string) => act(async () => {
        const input = getInput(container);
        Simulate.change(input, { target: { value } as unknown as EventTarget });
    });

    const submitForm = async (container) => act(async () => {
        const form = getForm(container);
        Simulate.submit(form);
    });

    // Field component debounces validate on change
    // jest@26 fake timers and lodash debounce are not friends
    // blur input to trigger a validation
    const blurInput = async (container) => act(async () => {
        const input = getInput(container);
        Simulate.blur(input);
    });

    const clickAddNewTag = async (container) => act(async () => {
        Simulate.click(getAddButton(container));
    });

    it('renders tags', () => {
        const tags = ['tag 1', 'tag 2'];
        const container = getComponent({ tags });

        const tagElements = scryRenderedDOMComponentsWithClass(container, 'mx_TagComposer_tag');
        expect(tagElements).toMatchSnapshot();
    });

    it('removes tags on remove button click', () => {
        const tags = ['tag 1', 'tag 2'];
        const onRemove = jest.fn();
        const container = getComponent({ tags, onRemove });

        const [tag1] = scryRenderedDOMComponentsWithClass(container, 'mx_TagComposer_tag');
        act(() => {
            Simulate.click(tag1.querySelector('[role=button]'));
        });
        expect(onRemove).toHaveBeenCalledWith('tag 1');
    });

    describe('form', () => {
        it('renders form', () => {
            const container = getComponent();

            const input = getInput(container);
            const addButton = getAddButton(container);

            expect(input).toBeTruthy();
            expect(addButton).toBeTruthy();
        });

        it('adds new tag when there is no validation', async () => {
            const onAdd = jest.fn();
            const container = getComponent({ onAdd });
            const newKeyword = "new keyword";

            await editNewTag(container, newKeyword);
            const input = getInput(container);
            expect(input.value).toEqual(newKeyword);

            await blurInput(container);

            await clickAddNewTag(container);

            expect(onAdd).toHaveBeenCalledWith(newKeyword);

            // clears input
            expect(input.value).toEqual("");
        });

        it('does not add tag with zero length', async () => {
            const onAdd = jest.fn();
            const container = getComponent({ onAdd });

            await editNewTag(container, "");
            await blurInput(container);

            await clickAddNewTag(container);

            expect(onAdd).not.toHaveBeenCalled();
        });

        describe('with validation', () => {
            it('adds tag on add button click when validation passes', async () => {
                const onAdd = jest.fn();
                const onValidate = jest.fn().mockResolvedValue({ valid: true });
                const container = getComponent({ onAdd, onValidate });
                const newKeyword = "new keyword";

                await editNewTag(container, newKeyword);
                await blurInput(container);

                await clickAddNewTag(container);

                expect(onAdd).toHaveBeenCalledWith(newKeyword);
            });

            it('disables add button when validation fails', async () => {
                const onAdd = jest.fn();
                const onValidate = jest.fn().mockResolvedValue({ valid: false });
                const container = getComponent({ onAdd, onValidate });
                const newKeyword = "new keyword";

                await editNewTag(container, newKeyword);
                await blurInput(container);

                expect(getAddButton(container).className.includes('mx_AccessibleButton_disabled')).toBeTruthy();
            });

            it('disables form submission when validation fails', async () => {
                const onAdd = jest.fn();
                const onValidate = jest.fn().mockResolvedValue({ valid: false });
                const container = getComponent({ onAdd, onValidate });
                const newKeyword = "new keyword";

                await editNewTag(container, newKeyword);
                await blurInput(container);
                await submitForm(container);
                expect(onAdd).not.toHaveBeenCalled();
            });
        });
    });
});
