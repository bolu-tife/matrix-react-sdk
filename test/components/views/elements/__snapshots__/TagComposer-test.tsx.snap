// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TagComposer /> renders tags 1`] = `
Array [
  <div
    class="mx_TagComposer_tag"
  >
    <span>
      tag 1
    </span>
    <div
      class="mx_AccessibleButton"
      role="button"
      tabindex="0"
    />
  </div>,
  <div
    class="mx_TagComposer_tag"
  >
    <span>
      tag 2
    </span>
    <div
      class="mx_AccessibleButton"
      role="button"
      tabindex="0"
    />
  </div>,
]
`;
